yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -94,25 +96,66 @@ export class AvmPersistableStateManager {
     return Promise.resolve(value);
   }
 
+  /**
+   * Write a note hash, trace the write.
+   * @param noteHash - the unsiloed note hash to write
+   */
   public writeNoteHash(noteHash: Fr) {
     this.trace.traceNewNoteHash(/*storageAddress*/ Fr.ZERO, noteHash);
   }
 
-  public async checkNullifierExists(storageAddress: Fr, nullifier: Fr) {
+  /**
+   * Check if a nullifier exists, trace the check.
+   * @param storageAddress - address of the contract that the nullifier is associated with
+   * @param nullifier - the unsiloed nullifier to check
+   * @returns exists - whether the nullifier exists in the nullifier set
+   */
+  public async checkNullifierExists(storageAddress: Fr, nullifier: Fr): Promise<boolean> {
     const [exists, isPending, leafIndex] = await this.nullifiers.checkExists(storageAddress, nullifier);
     this.trace.traceNullifierCheck(storageAddress, nullifier, exists, isPending, leafIndex);
     return Promise.resolve(exists);
   }
 
+  /**
+   * Write a nullifier to the nullifier set, trace the write.
+   * @param storageAddress - address of the contract that the nullifier is associated with
+   * @param nullifier - the unsiloed nullifier to write
+   */
   public async writeNullifier(storageAddress: Fr, nullifier: Fr) {
     // Cache pending nullifiers for later access
     await this.nullifiers.append(storageAddress, nullifier);
     // Trace all nullifier creations (even reverted ones)
     this.trace.traceNewNullifier(storageAddress, nullifier);
   }
 
-  public writeL1Message(message: Fr[]) {
-    this.newL1Messages.push(message);
+  /**
+   * Check if an L1 to L2 message exists, trace the check.
+   * @param msgHash - the message hash to check existence of
+   * @param msgLeafIndex - the message leaf index to use in the check
+   * @returns exists - whether the message exists in the L1 to L2 Messages tree
+   */
+  public async checkL1ToL2MessageExists(msgHash: Fr, msgLeafIndex: Fr): Promise<boolean> {
+    // TODO: we don't actually need to get the message, just need to check existence of the msg hash at the provided index
+    let exists = false;
+    try {
+      const gotMessage = await this.hostStorage.commitmentsDb.getL1ToL2Message(msgHash);
+      exists = gotMessage !== undefined && gotMessage.index == msgLeafIndex.toBigInt();
+    } catch {
```

Comment:
Should we add a function to the commitments DB `getL1ToL2MessageExists(msgHash, index)`? Only weirdness with adding things to the CommitmentsDB interface is that DBOracle extends it and SimulatorOracle implements it. Not sure we need it in SimulatorOracle?